### PR TITLE
[#153668369] passport saml

### PR DESCRIPTION
**Pull request by lussoluca**

### CODE REVIEW

File: app/src/server.js [hunk 1]
```diff
@@ -40,13 +40,15 @@ passport.use(
 passport.use(
   new SpidStrategy(
```
Comment (cloudify) [inline app/src/server.js]: for improved readability I would suggest to separate che creation of the strategy from the call to `passport.use`

**Comment by cloudify** [comment]
I would also suggest to NOT copy the code from the [spid-passport repo](https://github.com/italia/spid-passport/blob/master/index.js) in this project, instead we should fork the original project, improve it and send a PR (and until the PR gets merged, use our own fork).

**Comment by lussoluca** [comment]
I follow your suggestion and here you could find the PR for the spid-passport repo:

https://github.com/italia/spid-passport/pull/4

**Comment by lussoluca** [comment]
@cloudify at the moment the spid-passport isn't on npm, how do you want to proceed? The working code is [here](https://github.com/lussoluca/spid-passport) until the [PR](https://github.com/italia/spid-passport/pull/4) will be accepted

**Comment by cloudify** [comment]
I would either publish it on npm under a different namespace or even just reference the repo/commit from `package.json`

**Comment by lussoluca** [comment]
@cloudify I've updated the PR to point to git://github.com/lussoluca/spid-passport.git#master directly.

File: app/package.json [hunk 2]
```diff
@@ -27,12 +27,15 @@
...
     "dotenv": "^4.0.0",
     "express": "^4.16.0",
     "morgan": "^1.8.1",
     "ms-rest": "^2.2.7",
-    "passport": "^0.3.2",
-    "passport-http-bearer": "^1.0.1"
+    "passport": "^0.4.0",
+    "passport-strategy": "^1.0.0",
+    "passport-http-bearer": "^1.0.1",
+    "spid-passport": "git://github.com/lussoluca/spid-passport.git#master"
```
Comment (cloudify) [inline app/package.json]: I would set a specific commit to make the build reproducible

File: app/src/controllers/authenticationController.js [hunk 3]
```diff
@@ -21,32 +21,25 @@ export default class AuthenticationController {
...
   /**
+   * The Assertion consumer service.
    *
    * @param req
    * @param res
    */
-  sso(req: express$Request, res: express$Response) {
-    // Use the shibboleth session id as token.
-    const token = req.headers["shib-session-id"];
+  acs(req: express$Request, res: express$Response) {
```
Comment (cloudify) [inline app/src/controllers/authenticationController.js]: better name? `assertionConsumerService` perhaps?

File: app/src/server.js [hunk 4]
```diff
@@ -6,45 +6,50 @@
...
-      return done(null, false);
-    }
-  })
-);
+
+const port = process.env.PORT || 443;
+
+// Setup Passport.
+
+passport.use(tokenStrategy);
```
Comment (cloudify) [inline app/src/server.js]: pls add a comment to each line describing briefly what each passport and express plugin/middleware does

File: app/src/server.js [hunk 5]
```diff
@@ -6,45 +6,50 @@
...
+passport.use(spidStrategy);
+
+// Create and setup the Express app.
 
 const app = express();
 app.use(morgan("dev"));
```
Comment (cloudify) [inline app/src/server.js]: shouldn't take the name from the environment?

File: app/src/server.js [hunk 6]
```diff
@@ -6,45 +6,50 @@
...
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({ extended: false }));
 app.use(express.static("public"));
 app.use(passport.initialize());
 
-app.get("/sso", function(req: express$Request, res: express$Response) {
+// Setup routing.
+
+app.get("/login", passport.authenticate("spid", { session: false }));
```
Comment (cloudify) [inline app/src/server.js]: can `passport.authenticate` be called once? 

``` js

const spidAuth = passport.authenticate("spid", { session: false });

app.get("/login", spidAuth);
app.get("/acs", spidAuth, ...);

```

File: app/src/server.js [hunk 7]
```diff
@@ -6,45 +6,50 @@
...
+
+app.post("/acs", passport.authenticate("spid", { session: false }), function(
+  req: express$Request,
+  res: express$Response
+) {
   const controller = (container.resolve(
```
Comment (cloudify) [inline app/src/server.js]: shouldn't this be outside the endpoint cb? it's fair to assume the controller will not change from one request to another...

Comment (cloudify) [reply app/src/server.js]: if you resolve the controller once, you can then:

``` js
const acsController = ...
app.post("/acs", spidAth, acsController);
```

File: app/src/server.js [hunk 8]
```diff
@@ -59,7 +64,17 @@ app.get(
   }
 );
 
-const server = app.listen(port, function() {
+// Setup and start the HTTPS server.
+
+const key = fs.readFileSync("./certs/key.pem", "utf-8");
```
Comment (cloudify) [inline app/src/server.js]: these paths should be configurable via ENV vars

Comment (cloudify) [inline app/src/server.js]: these paths should be configurable via ENV vars (like `PORT`) - logging what's going on would be nice (i.e. loading certs and paths)

File: app/src/strategies/tokenStrategy.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
+// @flow
```
Comment (cloudify) [inline app/src/strategies/tokenStrategy.js]: briefly describe the purpose of this file

Comment (cloudify) [inline app/src/strategies/tokenStrategy.js]: briefly describe the purpose of this file

File: app/src/strategies/tokenStrategy.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+"use strict";
+
+import type { SessionStorageInterface } from "../services/sessionStorageInterface";
+import container from "../container";
+
+const Strategy = require("passport-http-bearer");
+
+const tokenStrategy = new Strategy(function(token, done) {
+  const sessionStorage = (container.resolve(
+    "sessionStorage"
```
Comment (cloudify) [inline app/src/strategies/tokenStrategy.js]: question about this IoC approach, how can we check that `container.resolve` gets called with valid identifiers? (e.g. catching typos in module name)

Comment (cloudify) [reply app/src/strategies/tokenStrategy.js]: also, how can we verify that `sessionStorage` has been instantiated before `tokenStrategy` is there a concept of dependency?

File: app/src/server.js (hunk 8 above)
Comment (lussoluca) [reply app/src/server.js]: Those files are mounted in the Docker container always in the same location, I don't think we need an ENV here

Comment (lussoluca) [reply app/src/server.js]: I'll open a story to introduce logging

Comment (cloudify) [reply app/src/server.js]: it's always convenient to being able to override a conf via env vars (see [12 factor apps](https://12factor.net/config)), for instance if you run the script locally, it's not big deal, you can just add:

``` js
const cert_key_path = process.env.CERT_KEY_PATH || "./certs/key.pem"; 
```

File: app/src/controllers/authenticationController.js (hunk 3 above)
Comment (lussoluca) [reply app/src/controllers/authenticationController.js]: Renamed

File: app/src/strategies/tokenStrategy.js (hunk 9 above)
Comment (lussoluca) [reply app/src/strategies/tokenStrategy.js]: Done

Comment (lussoluca) [reply app/src/strategies/tokenStrategy.js]: Done

File: app/src/container.js [hunk 11]
```diff
@@ -13,18 +19,22 @@ const container = awilix.createContainer({
   resolutionMode: awilix.ResolutionMode.CLASSIC
 });
 
+export const SESSION_STORAGE = "sessionStorage";
```
Comment (cloudify) [inline app/src/container.js]: DRY

``` js
export const SESSION_STORAGE = "sessionStorage";
container.registerClass({
  [SESSION_STORAGE]: [DummySessionStorage, { lifetime: awilix.Lifetime.SINGLETON }]
});

```

**Pull request merged**